File: algorithms/code/leetcode/lc050_pow_x_n/lc050_pow_x_n.py
#!/usr/bin/env python3
# https://leetcode.com/problems/powx-n/

import unittest


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1.0
        count = -int(n) if n < 0 else int(n)
        result = x
        power = 1
        powers = []
        while (power << 1) <= count:
            powers.append(result)
            result *= result
            power <<= 1
        previous = power >> 1
        for i in range(len(powers) - 1, -1, -1):
            repeat = (count - power) // previous
            value = powers[i]
            for _ in range(repeat):
                result *= value
            power += repeat * previous
            previous >>= 1
        result = 1.0 / result if n < 0 else result
        return result


class TestCode(unittest.TestCase):
    def test_smaller(self):
        x = 34.00515
        n = -3
        expected = x ** n
        self.assertEqual(expected, Solution().myPow(x, n), 0)

    def test_bigger(self):
        x = 0.00001
        n = 2147483647
        expected = x ** n
        self.assertEqual(expected, Solution().myPow(x, n), 0)

    def test_0(self):
        self.assertEqual(1.0, Solution().myPow(0, 0), 0)

File: algorithms/code/leetcode/lc050_pow_x_n/lc050_pow_x_n.rs
// https://leetcode.com/problems/powx-n/
// #medium

pub struct Solution;

impl Solution {
    pub fn my_pow(x: f64, n: i32) -> f64 {
        if n == 0 {
            return 1.0;
        }
        let count = if n < 0 { -(n as i64) } else { n as i64 };
        let mut result = x;
        let mut power = 1;
        let mut powers = Vec::new();
        while (power << 1) <= count {
            powers.push(result);
            result *= result;
            power <<= 1;
        }
        let mut previous = power >> 1;
        for i in (0..powers.len()).rev() {
            let repeat = (count - power) / previous;
            let value = powers[i];
            for _ in 0..repeat {
                result *= value;
            }
            power += repeat * previous;
            previous >>= 1;
        }
        if n < 0 { 1.0 / result } else { result }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_smaller() {
        let x = 34.00515;
        let n = -3;
        let expected = f64::powf(x, n as f64);
        assert_eq!(Solution::my_pow(x, n), expected);
    }

    #[test]
    fn test_bigger() {
        let x = 0.00001;
        let n = 2147483647;
        let expected = f64::powf(x, n as f64);
        assert_eq!(Solution::my_pow(x, n), expected);
    }

    #[test]
    fn test_0() {
        assert_eq!(Solution::my_pow(0.0, 0), 1.0);
    }
}
